Dev fp vs oop

Sumit-tech-joshi opened the pull request.

Sumit-tech-joshi committed:
Add comparison of FP and OOP

Sumit-tech-joshi committed:
Add other section for the comparison

Sumit-tech-joshi committed:
Merge branch 'dev-fp-vs-oop' of https://github.com/Sumit-tech-joshi/pattern-library into dev-fp-vs-oop

Sumit-tech-joshi committed:
revert code for the Functional Patterns vs. OOP

Sumit-tech-joshi committed:
Add comment on FP code

Sumit-tech-joshi commented:
Hi @ash-teach , The  #59  got closed due to merging conflicts, I have created new PR for all the changes requested in #59 in this PR, Please review and  let me know if anything needs to be added?

KaiKitKobold commented:
> Hi @ash-teach , The #59 got closed due to merging conflicts, I have created new PR for all the changes requested in #59 in this PR, Please review and let me know if anything needs to be added?

The pull request looks good! The hierarchy of the content looks good, and it meshes well with the other content already on the Read. Me.

nazrinzuwair commented:
Hi @Sumit-tech-joshi 

I like the way you have structured your content and its flow of information. It is very easy to understand, but I would suggest adding a couple more sections to the OOP Design Patterns and Functional Patterns. The descriptions are quite limited, and they miss most of the important outlines.

In my opinion, I prefer having the coding example of the Singleton within the overview itself, without redirecting to a link. Alternatively, you can provide a short code snippet so it is easier to make comparisons between the OOP Design Patterns and Functional Patterns.

Additionally, you can stick more to JavaScript and deliver some information on how it is used in OOP Design Patterns and Functional Patterns. Programming language implementations differ in each, so a brief overview may be helpful.

I like the contrast and comparison you have made, but if you present it within a table-like structure, it will be easier for the user to grasp its idea.

Sumit-tech-joshi committed:
Add table structure for contrast

Sumit-tech-joshi commented:
Hello @nazrinzuwair,

Previously, in PR #59, I included an example of the singleton pattern. However, based on the feedback from @ash-teach, I have removed the code and replaced it with a link to the code. 

Also both of the implementation contains code from the javascript the OOP design has map function widely use in javascript which i explained by the comments.

For better understanding the contrast i have added table structure as you suggested.  Please see the [Commit](https://github.com/nic-dgl104-winter-2024/pattern-library/pull/73/commits/e14fb6f2b56f3c2506debf2d4b7e5b8faf6ffcd0)

Thank you for putting your concerns.

Sumit-tech-joshi commented:
@ash-teach Could you please review this again after changes.

ash-teach commented on README.md:
````diff
@@ -153,7 +153,59 @@ Michaelstonis. (2022, November 4). _Model-View-ViewModel - .NET._ Microsoft Lear
+One common pattern in FP is the Higher-Order Function pattern. Higher-order functions take other functions as arguments or return them, enabling composition and abstraction.
+
+```
+JavaScript
````
'javascript' (all lowercase) should appear on the same line as the triple backticks, like this:
```
```javascript
```

Sumit-tech-joshi replied:
Change Done.

ash-teach commented on README.md:
```diff
@@ -153,7 +153,59 @@ Michaelstonis. (2022, November 4). _Model-View-ViewModel - .NET._ Microsoft Lear
+**Composition vs. Inheritance:** Functional programming encourages composition over inheritance. Instead of relying on class hierarchies, functions are composed to achieve complex behaviors.
+
+**Pattern Matching:** FP languages like Haskell or Scala offer pattern matching, allowing elegant handling of different cases. This contrasts with traditional switch-case or if-else statements in OOP.
+
```
I like the table below, however it makes the statements above redundant. Perhaps as an improvement you could swap out the descriptions above with actual definitions for each. I think in this case it's fine to reproduce (i.e. quote) definitions directly from a reputable source - as long as you make clear they're quotes. For example, you could use the following for pattern matching:

> **Pattern Matching**: "the act of checking a given sequence of tokens for the presence of the constituents of some pattern" ([Wikipedia -Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching)).

For the purposes of keeping things simple, include no more than one or two sentences for each of immutability, composition, and inheritance (you can use pattern matching as I've provided). Don't worry about explaining complicated terms, but do make sure you get the right definitions (i.e. object composition in programming - not musical composition!) And you can use Wikipedia as a source.

Sumit-tech-joshi replied:
Changes done

ash-teach replied:
I actually meant the other way around. I think the table was nice because it distinguished between functional and oop in an easy-to-read way. Do you mind updating one more time?:

- [ ] Definitions moved above the table and _replacing_ lines 187 - 191.
- [ ] Put the table back as you had it in the previous commit.
- [ ] I would also separate the definitions of composition and inheritance for clarity. The quote you sourced does explain the distinction, but it does really say what each one is. Wikipedia is a good source for both.

Sumit-tech-joshi replied:
Changes done.

ash-teach commented:
Looks great, @Sumit-tech-joshi . Thanks for resubmitting the PR. I have two reasonably small change requests. If you can address these then I'll merge.

nazrinzuwair commented:
Looks great now @Sumit-tech-joshi

Sumit-tech-joshi committed:
Add javascript change for code snippet

Sumit-tech-joshi committed:
Resolve PR comment for table content change.

Sumit-tech-joshi commented:
@ash-teach I have done the changes as requested, could you please have a look again and let me know if further more changes are required.

ash-teach commented:
Thanks for the work, @Sumit-tech-joshi  - it looks like we had a miscommunication though. I've put more details in the comment in the change request. Let me know if you need any more details!

Sumit-tech-joshi committed:
Done PR comments changes

Sumit-tech-joshi commented:
@ash-teach , Thank you for the comment explanation, I have done the changes as requests, Hope this time got it correct  🤞.
